sdks/go/pkg/beam/io/synthetic/source_test.go

```diff
@@ -162,6 +164,83 @@ func TestSourceConfig_BuildFromJSON(t *testing.T) {
 	}
 }
 
+// TestSourceConfig_NumHotKeys tests that setting the number of hot keys
+// for a synthetic source works correctly.
+func TestSourceConfigBuilder_NumHotKeys(t *testing.T) {
+	tests := []struct {
+		elms    int
+		hotKeys int
+	}{
+		{elms: 15, hotKeys: 2},
+		{elms: 30, hotKeys: 10},
+		{elms: 50, hotKeys: 25},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(fmt.Sprintf("(elm = %v)", test.hotKeys), func(t *testing.T) {
+			dfn := sourceFn{}
+			cfg := DefaultSourceConfig()
+			cfg.NumElements(test.elms)
+			cfg.HotKeyFraction(1.0)
+			cfg.NumHotKeys(test.hotKeys)
+
+			keys, _, err := simulateSourceFn(t, &dfn, cfg.Build())
+			if err != nil {
+				t.Errorf("Failure processing sourceFn: %v", err)
+			}
+
+			type val struct {
+				num      int
+				isHotKey bool
+			}
+
+			m := make(map[string]val)
```

Comment:
I think this whole logic could be simplified. What we need to do is to count distinct elements in the `keys` slice. So it can look something like this:

```go
counter := make(map[byte]int) // no need to convert to string
for _, key := range keys {
    counter[key]++
}
```
After that you can use `len(keys)` and `len(counter)` in assertions.

Reply:
Thank you for help! 

@kamilwu wrote:
```golang
counter := make(map[byte]int) // no need to convert to string
```

I think it should be byte array
```golang
counter := make(map[[]byte]int) 
```
And you can't use byte array in map.

```golang
for _, key := range keys {
    counter[key]++
}
```
You're right. In the last commit I removed the struct and used direct counter, also I added support for testing HotKeyFraction. However I'm a bit worried that testing HotKeyFraction could be Flaky, what do you think?

Reply:
Yes, you're right, it should be a byte array. We can't use arrays/slices as a key, but we can convert `[]byte` to `string`. You've already done this, so this is good!

> However I'm a bit worried that testing HotKeyFraction could be Flaky, what do you think?

Yes, there is some probability that this would happen unless we mock the generator. Let's set `HotKeyFraction` to 1.0. Testing only `NumHotKeys` is fine for me.

Reply:
@kamilwu wrote:
```Let's set HotKeyFraction to 1.0. Testing only NumHotKeys is fine for me.```

I commented test cases with HotKeyFraction < 1.0 and left the logic behind it. Is that ok?

Reply:
I left the comment below.